Gstreamer plugin

edersondisouza opened the pull request:
A simple gstreamer plugin interface to aid testing AVB-SH video features. Right now, it's only MPEG-TS compatible.

Note that this PR is against branch `next`, so that it can be merged without risks on `master`. When time is due, anything on `next` can go into `master`.

edersondisouza committed:
A gstreamer plugin sample: avbvideosink element

To aid visualisation of the videos being sent through AVB-SH, one can
use a GStreamer interface, so that packets can follow a GStreamer
pipeline.

This patch adds a gstreamer sink element, that can get MPEG-TS packets
and send them to the AVB-SH AvbVideoBridge - which gets the packets and
send them down AVB-SH stack.

This element, `avbvideosink`, has a property "stream-name", associated
to the "IPC name" property set on AVB-SH configuration. That way, the
element can connect to a specific AVB-SH stream.

An example of using this element could be the following gstreamer
pipeline, on the talker:

$ LD_LIBRARY_PATH=$AVB_DEPS/lib/:. GST_PLUGIN_PATH=./build gst-launch-1.0 videotestsrc pattern=ball ! clockoverlay ! vaapih264enc ! mpegtsmux ! avbvideosink

This will send the video produced by `videotestsrc` (with the clock
overlay added by `clockoverlay`) to the h264 encoder, the MPEG-TS muxer
and finally to the avbvideosink, which will send it via network using
AVB-SH stack.

edersondisouza committed:
gstreamer plugin: A clock to represent the gPTP clock

GStreamer can work with multiple clocks, with different usages. This
patch provides a clock based on the gPTP clock used by AVB-SH. It will
be useful to correctly sync the presentation timer for the buffers
generated by the future source element.

Note that right now, the clock device is hardcoded as "/dev/ptp0", but
there are ways to get the right clock device from the network. This is
left for a future patch, though.

edersondisouza committed:
gstreamer plugin: avbvideosrc element

This patch adds a gstreamer source element, that can get MPEG-TS packets
from AVB-SH AvbVideoBridge and send them to a gstreamer pipeline.

This element, `avbvideosrc`, has a property "stream-name", associated
to the "IPC name" property set on AVB-SH configuration. That way, the
element can connect to a specific AVB-SH stream.

An example of using this element could be the following gstreamer
pipeline, on the talker:

$ LD_LIBRARY_PATH=$AVB_DEPS/lib/:. GST_PLUGIN_PATH=./build gst-launch-1.0 avbvideosource ! typefind ! tsdemux ! decodebin ! videoconvert ! clockoverlay halignment=right ! xvimagesink display=:0

This will get the video from the network via AVB-SH stack, demux it from
the MPEG-TS, decode it, add a clock overlay and display it on X display
0.

edersondisouza committed:
Configuration reference: Add a new MPEG-TS stream

So one can tinker with more than one stream - for instance, using the
gstreamer elements interface.

edersondisouza committed:
README: Add information about using the gstreamer plugins samples

Information about sending/receiving video via AVB-SH gstreamer interface
- and how to run more than one video stream.

vcgomes commented on private/src/gst/gstavbvideosink.cpp:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    ret = GST_ELEMENT_CLASS(gst_avbvideosink_parent_class)->change_state(element, transition);
+    if (ret == GST_STATE_CHANGE_FAILURE) {
```
Is there any chance of `avbvideosink->sender` leaking in this case?

vcgomes commented on private/src/gst/gstavbvideosrc.cpp:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    ret = GST_ELEMENT_CLASS(gst_avbvideosrc_parent_class)->change_state(element, transition);
+    if (ret == GST_STATE_CHANGE_FAILURE) {
```
I think the question that I made in the previous patch applies here: is is possible to leak the `receiver` here?

edersondisouza replied:
Nice question. I'll check if I have some assurances here.

keerockl replied:
I went through code change with Ederson and it looks good to me.

vcgomes commented on private/src/gst/gstavbvideosrc.cpp:
```diff
@@ -0,0 +1,246 @@
+        GST_PAD_ALWAYS,
+        GST_STATIC_CAPS(
+            "video/mpegts,"
+            "  packetsize = (int) 188,"
```
Please add a `#define` for 188. I am seeing it repeated in a few places.

edersondisouza replied:
Ok. But here I would need a second define as string, or some magic preprocessing stuff, like two levels of macros.

edersondisouza replied:
(Or be a rebel and let here as is. `C`, why won't you help me?)

vcgomes replied:
Another solution would be to build that capabilities string more dynamically, is there a not too awkward way of doing that?

edersondisouza replied:
Oh, https://developer.gnome.org/glib/stable/glib-Miscellaneous-Macros.html#G-STRINGIFY:CAPS came to the rescue =D

vcgomes replied:
I was thinking about not using GST_STATIC_CAPS and using something else. But this is getting too complicated, I agree. I think the two `#defines` (one for the value, the other for the string) are fine for now. What do you think?

vcgomes commented on private/src/gst/gstavbvideosrc.cpp:
```diff
@@ -0,0 +1,246 @@
+        GST_WARNING_OBJECT(avbvideosrc, "Packet may contain incomplete data. Size %zd is not a multiple of %u", packet->size, expected_size);
+    }
+
+    for (unsigned j = 0; packet->size - j >= expected_size; j += expected_size) {
```
I think the stop condition of the `for` is wrong, as you are dealing with `unsigned int`s and `packet->size` is a `size_t` (also unsigned), you may not detect the overflow case. Does that make sense?

edersondisouza replied:
You mean,  `j` overflowing before consuming all `packet->size`? That would be a problem sure. But then I'm at a loss to understand if the problem is `j` not being `size_t` or something else at the stop condition...

vcgomes replied:
For example, in that (error) case when `packet->size` is not a multiple of `expected_size`: `packet->size - j` will get promoted to an unsigned quantity, which will make the loop to iterate quite a few times more than you are expecting.

The problem I am seeing is that you won't be able to detect when `packet->size -j` becomes negative.

edersondisouza replied:
Why would it become negative?
`packet->size` is constant, `j` increases by `expected_size` each iteration. The first loop in which `packet->size -j` is smaller than `expected_size` will end the loop (if it is a multiple, the number will be 0, if not, it will be some `x` such as `0 < x < expected_size`). So I don't see it becoming negative. 
The only issue here is `j` not being `size_t`, so it could overflow (highly unlikely, but better safe than sorry =D ).
Of course, if you can show me a counterexample, I'll see new things.

vcgomes replied:
What I was thinking was the case that `packet->size` is smaller than 188. But reading the code again, the code is safe.

Sorry for the noise :-)

edersondisouza commented:
On Wed, 2018-10-03 at 15:30 -0700, Vinicius Gomes wrote:
> @vcgomes commented on this pull request.
> In private/src/gst/gstavbvideosrc.cpp:
> > +
> +enum
> +{
> +    PROP_0,
> +    PROP_STREAM_NAME
> +};
> +
> +/* pad templates */
> +
> +static GstStaticPadTemplate gst_avbvideosrc_src_template =
> +GST_STATIC_PAD_TEMPLATE("src",
> +        GST_PAD_SRC,
> +        GST_PAD_ALWAYS,
> +        GST_STATIC_CAPS(
> +            "video/mpegts,"
> +            "  packetsize = (int) 188,"
> Another solution would be to build that capabilities string more dynamically, is there a not too awkward way of doing that?

Maybe `C++ std::string` can help. I'll check it.

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

edersondisouza commented:
v2:
 - Fixed possible leaks of `ias_avbvideobridge_sender` and  `ias_avbvideobridge_receiver`.
 - Fixed possible overflow of `unsigned int` variable
 - Removed spurious log message
 - Defined constant for `188` MPEG-TS packet size and others

vcgomes commented:
Looks good from my side.

And sorry for taking this long.

edersondisouza commented:
Could someone merge this PR on the next branch, so I can close it?
